### PR TITLE
[Profiler] add idle time and cuda kernel queue depth computation

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79805

Test Plan:
Add tests in test_profiler.py
Check results with all the torchbench models

### CODE REVIEW

File: torch/profiler/_utils.py [hunk 1]
```diff
@@ -18,32 +42,214 @@ def __eq__(self, other):
...
+        cuda_kernel_events = [
+            event for event in event_list
+            if is_cuda_launch_kernel(event) or is_cuda_kernel(event)
+        ]
+        cuda_kernel_events.sort(key=lambda e: e.start_us())
+
+        for curr_event, next_event in zip(cuda_kernel_events,
+                                          cuda_kernel_events[1:]):
+            if (is_cuda_launch_kernel(curr_event)):
+                if (curr_qd == 0):
```
Comment (robieta) [inline torch/profiler/_utils.py:119]: This method of counting strikes me as rather brittle. If you have a single launch where we don't capture the kernel then `curr_qd` will never go to zero. And by contrast, if there are more kernels than launches (recall that kernels can call other kernels in CUDA) curr_qd will be negative and all sorts of craziness will happen.

A potentially more robust solution would be to keep track of blocks of kernels (`List[List[_KinetoEvent]]`) and make a new block when a kernel launch had a start time greater than the end time of the last event in the current block.

File: torch/profiler/_utils.py [hunk 2]
```diff
@@ -18,32 +42,214 @@ def __eq__(self, other):
...
+                         (next_event.start_us() + next_event.duration_us()) *
+                         1000, curr_qd))
+
+        event_list = [e.event for e in self.metrics.keys()]
+        for event in event_list:
+            self.metrics[EventKey(event)].idle_time_ns = EventKey(
+                event).intervals_overlap(idle_interval)
+        return qd_list
+
+    def rank_events(self, length):
```
Comment (robieta) [inline torch/profiler/_utils.py:141]: This is really a second PR, so I'm going to defer and recommend that you split this out.

File: test/test_profiler.py [hunk 3]
```diff
@@ -1127,6 +1127,28 @@ def test_utils_compute_self_time(self):
                     for child in event_key.event.children
                 ]))
 
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA is required")
```
Comment (robieta) [inline test/test_profiler.py:1130]: This doesn't really test the algorithm; it just checks that it runs and produces values that are in bounds. I would suggest that you break it into the following two tests:

A real test with very coarsely controlled timings. The "real" test should be something like:
```
x = torch.ones((4096, 4096), device="cuda")
for _ in range(5):
  y = torch.mm(x, x)
torch.cuda.synchronize()
for _ in range(3):
  x[0] += 1
  time.sleep(0.1)
```
We want a region well into the pipelined queue domain and one well into the dispatch bound domain. The code above is just off the top of my head, so you may need to tune it. You should then be able to write more quantitative tests than just `>= 0.0` and `<= 1.0`

A test where you mock `_KinetoEvent` and assert that the algorithm produces exact values.

**Review by robieta — CHANGES_REQUESTED** [review]

File: torch/profiler/_utils.py (hunk 1 above)
Comment (davidchencsl) [reply torch/profiler/_utils.py:119]: I am not quite getting your solution here, could you elaborate on that a little bit?

Comment (robieta) [reply torch/profiler/_utils.py:119]: After thinking a bit I decided the nested lists probably aren't needed. But the general procedure is the same.

Suppose we partition events into kernel launches and kernels, and maintain a mapping from kernel to launch. (If not directly assignable by correlation ID, check previous kernels until we find one that is.) We can then walk the cpu events and get the gpu kernel currently executing when the event starts (if any) by advancing an index into the GPU kernels. We know the index of the kernel launched by the current event (or the last launch if the current event is not a cuda kernel launch), and the difference in those indices is the queue depth at that op.

**Pull request closed without merging**